human-tracking-and-pose-detection/main.py

```diff
@@ -0,0 +1,217 @@
+import argparse
+import cv2
+import math
+import numpy as np
+import tensorflow as tf
+import tensorflow_hub as hub
+import time
+parser_args = tuple()
+
+
+def parse_arguments():
+    """
+    handle command line arguments
+    """
+    parser = argparse.ArgumentParser(description='Example command-line parser')
+    parser.add_argument(
+        '-v',
+        '--verbose',
+        action='store_true',
+        help='Enable print_verbose_only outputs')
+    parser.add_argument(
+        '-k',
+        '--display_keypoint',
+        action='store_true',
+        help='Enable display of keypoints')
+    parser.add_argument(
+        '-i',
+        '--display_image',
+        action='store_true',
+        help='Enable display of image and keypoints')
+    return parser.parse_args()
+
+
+def print_verbose_only(*args, **kwargs):
+    """
+    print only if verbose==True
+    """
+    if parser_args.verbose:
+        print(*args, **kwargs)
+
+
+def print_fps(func):
+    """
+    function decorator for printing fps
+    """
+    def wrapper(*args, **kwargs):
+        prev_time = time.time()
+        result = func(*args, **kwargs)
+        cur_time = time.time()
+        if cur_time - prev_time > (0.000001):
+            fps = 1.0 / (cur_time - prev_time)
+            print_verbose_only('FPS: {:.2f}'.format(fps))
+        return result
+    return wrapper
+
+
+def movenet(input_image, model):
+    """
+    movenet model:
+    Gets input image and outputs array of keypoints with certainty score
+    downloaded from #https://tfhub.dev/google/movenet/multipose/lightning/1
+    """
+    # SavedModel format expects tensor type of int32.
+    input_image = tf.cast(input_image, dtype=tf.int32)
+    outputs = model(input_image)  # Output is a [1, 6, 56] tensor.
+
+    keypoints = outputs['output_0'].numpy()
+
+    threshold = 0.3
+    count_of_people = np.sum(keypoints[0, :, -1] > threshold)
+    print_verbose_only("count_of_people", count_of_people)
+    return keypoints[:, :, :51].reshape((6, 17, 3))[0]
```

Comment:
What are these numbers?

Reply:
i will put this comment in code:
there are 6 people each with 17 body parts.
each body part have [x, y, certainty_score], so there are 3*17=51 numbers per person
The first 17 * 3 elements are the keypoint locations and scores in the format: [y_0, x_0, s_0, y_1, x_1, s_1, …, y_16, x_16, s_16], where y_i, x_i, s_i are the yx-coordinates (normalized to image frame, e.g. range in [0.0, 1.0]) and confidence scores of the i-th joint correspondingly. The order of the 17 keypoint joints is: [nose, left eye, right eye, left ear, right ear, left shoulder, right shoulder, left elbow, right elbow, left wrist, right wrist, left hip, right hip, left knee, right knee, left ankle, right ankle]. 

threshold is the minimum certain score needed for a point to be not ignored
the [0] is extra code that i have just removed